File: DAY 4/solutions.java/5. Longest Palindromic Substring.java
class Solution {

    public String longestPalindrome(String s) {

        if (s.length() < 2) {
            return s;
        }

        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {

            int[] odd = expand(s, i, i);
            int[] even = expand(s, i, i + 1);

            if (odd[1] - odd[0] > end - start) {
                start = odd[0];
                end = odd[1];
            }

            if (even[1] - even[0] > end - start) {
                start = even[0];
                end = even[1];
            }
        }

        return s.substring(start, end + 1);
    }

    private int[] expand(String s, int left, int right) {

        while (left >= 0 &&
               right < s.length() &&
               s.charAt(left) == s.charAt(right)) {

            left--;
            right++;
        }

        return new int[]{left + 1, right - 1};
    }
}

File: DAY 4/solution.py/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:

        if len(s) < 2:
            return s

        start = 0
        end = 0

        for i in range(len(s)):

            left, right = self.expand(s, i, i)
            left2, right2 = self.expand(s, i, i + 1)

            if right - left > end - start:
                start = left
                end = right

            if right2 - left2 > end - start:
                start = left2
                end = right2

        return s[start:end + 1]

    def expand(self, s, left, right):

        while left >= 0 and right < len(s) and s[left] == s[right]:

            left -= 1
            right += 1

        return left + 1, right - 1